docs/proposals/lvmd-within-topolvm-node.md

```diff
@@ -0,0 +1,404 @@
+// It also includes a watch for redirecting the vg watch to the local caller.
+type localCaller struct {
+	lvServiceServer proto.LVServiceServer
+	vgServiceServer proto.VGServiceServer
+	vgWatch         *localWatch
+}
+
+// localWatch is a local implementation of the WatchClient and WatchServer that is used by vgService.
+// it uses an unbound blocking channel to send and receive messages.
+type localWatch struct {
+	proto.VGService_WatchClient
+	proto.VGService_WatchServer
+	watch chan any
+	ctx   context.Context
+}
+
+func (l *localWatch) Recv() (*proto.WatchResponse, error) {
+	m := new(proto.WatchResponse)
+	if err := l.RecvMsg(m); err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+func (l *localWatch) Send(m *proto.WatchResponse) error {
+	return l.SendMsg(m)
+}
+
+func (l *localWatch) Header() (metadata.MD, error) {
+	return map[string][]string{}, nil
+}
+
+func (l *localWatch) Trailer() metadata.MD {
+	return map[string][]string{}
+}
+
+func (l *localWatch) CloseSend() error {
+	return nil
+}
+
+func (l *localWatch) Context() context.Context {
+	return l.ctx
+}
+
+func (l *localWatch) SendMsg(m interface{}) error {
+	l.watch <- m
+	return nil
+}
+
+func (l *localWatch) RecvMsg(m interface{}) error {
+	var ok bool
+	received, ok := <-l.watch
+	if !ok {
+		return status.Error(codes.Aborted, "watch closed")
+	}
+	receivedMsg, ok := received.(gproto.Message)
+	if !ok {
+		return status.Error(codes.Internal, "did not receive Message")
+	}
+	mMsg, ok := m.(gproto.Message)
+	if !ok {
+		return status.Error(codes.Internal, "pointer passed is not a Message")
+	}
+	gproto.Merge(mMsg, receivedMsg)
+	return nil
+}
+
+func (l *localCaller) Watch(_ context.Context, _ *proto.Empty, _ ...grpc.CallOption) (proto.VGService_WatchClient, error) {
+	return l.vgWatch, nil
+}
+
+func (l *localCaller) CreateLV(ctx context.Context, in *proto.CreateLVRequest, _ ...grpc.CallOption) (*proto.CreateLVResponse, error) {
+	return l.lvServiceServer.CreateLV(ctx, in)
+}
+
+func (l *localCaller) RemoveLV(ctx context.Context, in *proto.RemoveLVRequest, _ ...grpc.CallOption) (*proto.Empty, error) {
+	return l.lvServiceServer.RemoveLV(ctx, in)
+}
+
+func (l *localCaller) ResizeLV(ctx context.Context, in *proto.ResizeLVRequest, _ ...grpc.CallOption) (*proto.Empty, error) {
+	return l.lvServiceServer.ResizeLV(ctx, in)
+}
+
+func (l *localCaller) CreateLVSnapshot(ctx context.Context, in *proto.CreateLVSnapshotRequest, _ ...grpc.CallOption) (*proto.CreateLVSnapshotResponse, error) {
+	return l.lvServiceServer.CreateLVSnapshot(ctx, in)
+}
+
+func (l *localCaller) GetLVList(ctx context.Context, in *proto.GetLVListRequest, _ ...grpc.CallOption) (*proto.GetLVListResponse, error) {
+	return l.vgServiceServer.GetLVList(ctx, in)
+}
+
+func (l *localCaller) GetFreeBytes(ctx context.Context, in *proto.GetFreeBytesRequest, _ ...grpc.CallOption) (*proto.GetFreeBytesResponse, error) {
+	return l.vgServiceServer.GetFreeBytes(ctx, in)
+}
```

Comment:
How about replacing code with simple flows as in the following proposal?

https://github.com/topolvm/topolvm/blob/main/docs/proposals/thin-snapshots-restore.md#flow-of-operations-for-snapshot-creation

I believe it can reduce the review cost since we can focus on the design rather than the detail which will be done in code review.